File: app/src/main/java/com/battlelancer/seriesguide/util/TmdbTools.java
package com.battlelancer.seriesguide.util;

import android.content.Context;
import com.battlelancer.seriesguide.settings.TmdbSettings;
import com.uwetrottmann.tmdb2.entities.Genre;
import java.util.List;

public class TmdbTools {

    public enum ProfileImageSize {

        W45("w45"),
        W185("w185"),
        H632("h632"),
        ORIGINAL("original");

        private final String value;

        ProfileImageSize(String value) {
            this.value = value;
        }

        @Override
        public String toString() {
            return value;
        }
    }

    private static final String BASE_URL = "https://www.themoviedb.org/";
    private static final String PATH_TV = "tv/";
    private static final String PATH_MOVIES = "movie/";
    private static final String PATH_PERSON = "person/";

    /**
     * Tries to display the TMDb website of the given movie through a view intent.
     */
    public static void openTmdbMovie(Context context, int movieTmdbId) {
        openTmdbUrl(context, buildMovieUrl(movieTmdbId));
    }

    /**
     * Tries to display the TMDb website of the given person through a view intent.
     */
    public static void openTmdbPerson(Context context, int personTmdbId) {
        openTmdbUrl(context, buildPersonUrl(personTmdbId));
    }

    private static void openTmdbUrl(Context context, String url) {
        Utils.launchWebsite(context, url);
    }

    public static String buildEpisodeUrl(int showTmdbId, int season, int episode) {
        return BASE_URL + PATH_TV + showTmdbId + "/season/" + season + "/episode/" + episode;
    }

    public static String buildShowUrl(int showTmdbId) {
        return BASE_URL + PATH_TV + showTmdbId;
    }

    public static String buildMovieUrl(int movieTmdbId) {
        return BASE_URL + PATH_MOVIES + movieTmdbId;
    }

    public static String buildPersonUrl(int personTmdbId) {
        return BASE_URL + PATH_PERSON + personTmdbId;
    }

    /**
     * Build url to a profile image using the given size spec and current TMDb image url (see {@link
     * com.battlelancer.seriesguide.settings.TmdbSettings#getImageBaseUrl(android.content.Context)}.
     */
    public static String buildProfileImageUrl(Context context, String path, ProfileImageSize size) {
        return TmdbSettings.getImageBaseUrl(context) + size + path;
    }

    /**
     * Builds a string listing all given genres by name, separated by comma.
     */
    public static String buildGenresString(List<Genre> genres) {
        if (genres == null || genres.isEmpty()) {
            return null;
        }
        StringBuilder genresString = new StringBuilder();
        for (int i = 0; i < genres.size(); i++) {
            Genre genre = genres.get(i);
            genresString.append(genre.name);
            if (i + 1 < genres.size()) {
                genresString.append(", ");
            }
        }
        return genresString.toString();
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/tmdbapi/TmdbTools.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright © 2014 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.tmdbapi

import android.content.Context
import com.battlelancer.seriesguide.settings.DisplaySettings.isVeryHighDensityScreen
import com.battlelancer.seriesguide.settings.TmdbSettings
import com.uwetrottmann.tmdb2.entities.Genre

object TmdbTools {

    private const val IMAGE_SIZE_SPEC_ORIGINAL = "original"
    const val POSTER_SIZE_SPEC_W154 = "w154"
    const val POSTER_SIZE_SPEC_W342 = "w342"
    /**
     * As of 2024-11:
     *
     * - w300 (300 × 169 px) JPEG is around 9-16 kB
     * - w780 (780 × 439 px) JPEG is around 30-70 kB
     * - w1280 (1280 × 720 px) JPEG is around 60-140 KB
     *
     * Samples:
     *
     * - https://image.tmdb.org/t/p/original/8Tvnx22rzhwArofFPhmcfaBvgjN.jpg (smallest)
     * - https://image.tmdb.org/t/p/original/j4WEC9Jh4AyXF8ynpX3pz633tse.jpg
     * - https://image.tmdb.org/t/p/original/5Bh7EE3p6OOS0NzH22AE0N7DYO8.jpg (largest)
     */
    const val BACKDROP_SMALL_SIZE_SPEC = "w780"

    enum class ProfileImageSize(
        private val value: String
    ) {
        W45("w45"),
        W185("w185"),
        H632("h632"),
        ORIGINAL(IMAGE_SIZE_SPEC_ORIGINAL);

        override fun toString(): String {
            return value
        }
    }

    private const val BASE_URL = "https://www.themoviedb.org"
    private const val PATH_TV = "tv"
    private const val PATH_MOVIES = "movie"
    private const val PATH_PERSON = "person"

    @JvmStatic
    fun buildEpisodeUrl(showTmdbId: Int, season: Int, episode: Int): String {
        return "$BASE_URL/$PATH_TV/$showTmdbId/season/$season/episode/$episode"
    }

    @JvmStatic
    fun buildShowUrl(showTmdbId: Int): String {
        return "$BASE_URL/$PATH_TV/$showTmdbId"
    }

    @JvmStatic
    fun buildMovieUrl(movieTmdbId: Int): String {
        return "$BASE_URL/$PATH_MOVIES/$movieTmdbId"
    }

    fun buildPersonUrl(personTmdbId: Int): String {
        return "$BASE_URL/$PATH_PERSON/$personTmdbId"
    }

    fun buildMovieReleaseDatesUrl(movieTmdbId: Int): String {
        return "${buildMovieUrl(movieTmdbId)}/releases"
    }

    /**
     * Returns base image URL based on screen density.
     */
    fun getPosterBaseUrl(context: Context): String {
        return if (isVeryHighDensityScreen(context)) {
            TmdbSettings.getImageBaseUrl(context) + POSTER_SIZE_SPEC_W342
        } else {
            TmdbSettings.getImageBaseUrl(context) + POSTER_SIZE_SPEC_W154
        }
    }

    fun buildLargePosterUrl(context: Context, path: String): String =
        TmdbSettings.getImageBaseUrl(context) + POSTER_SIZE_SPEC_W342 + path

    fun buildOriginalSizeImageUrl(context: Context, path: String): String =
        TmdbSettings.getImageBaseUrl(context) + IMAGE_SIZE_SPEC_ORIGINAL + path

    fun buildBackdropUrl(context: Context, path: String): String =
        TmdbSettings.getImageBaseUrl(context) + BACKDROP_SMALL_SIZE_SPEC + path

    /**
     * Build URL to a profile image using the given size spec and current TMDB image url
     * (see [TmdbSettings.getImageBaseUrl]).
     */
    fun buildProfileImageUrl(context: Context, path: String?, size: ProfileImageSize): String? {
        return if (path == null) {
            null
        } else {
            TmdbSettings.getImageBaseUrl(context) + size + path
        }
    }

    /**
     * Builds a string listing all given genres by name, separated by comma.
     */
    fun buildGenresString(genres: List<Genre>?): String? {
        if (genres.isNullOrEmpty()) {
            return null
        }
        val genresString = StringBuilder()
        for (i in genres.indices) {
            val genre = genres[i]
            genresString.append(genre.name)
            if (i + 1 < genres.size) {
                genresString.append(", ")
            }
        }
        return genresString.toString()
    }
}
